Generate Board Information metadata in CI

junwoo091400 opened the pull request:
## Describe problem solved by this pull request
There were multiple instances where the QGC didn't have the fully updated list of bootloader IDs for the boards supported by PX4, which led to **users not being able to flash their boards as QGC didn't recognize them** (and didn't know which firmware was needed to be flashed)

For example, a user reported problem here: https://github.com/PX4/PX4-Autopilot/issues/19432 and it was manually fixed in QGC side here: https://github.com/mavlink/qgroundcontrol/pull/10252. **This is not scalable**, and is a lot of manual labor of copying the data from the PX4's side to QGC side (although it's not too intensive).

### Key Improvements
1. QGC being able to automatically detect new boards by it's board ID, when the board is added in PX4 side
2. QGC being able to auto-connect new boards by it's vendor & product ID (and name), when a new board is added in PX4 side

## Describe your solution
![PX4_Board_Information_Json](https://user-images.githubusercontent.com/23277211/176460048-1e7a2e4e-8f80-4004-b12b-441a266d98cb.png)

The aim of this PR is to automatically generate the **board information JSON file** that includes all the information needed to flash / build / download target builds, which QGC can then download and update the supported boards list to the latest data.

### Update to QGC Repository
First, the JSON file will get updated to the QGC Repository as a Pull Request or a direct commit like [how the Parameter / Airframe Metadata is getting synced](https://github.com/mavlink/qgroundcontrol/commit/1fa8485f7fa12eb19b3fc07f2cffe4bc2d3bb8c4) via Jenkins CI. This will make sure that QGC will have the latest Board information and will always fully support all the boards for flashing :+1: 

### Update to Local QGC
Second, the JSON file will be uploaded to a S3 server, for example to a place like: https://px4-travis.s3.amazonaws.com/Firmware/master/_general/board_info.json. This will then get pulled to local QGC installations so that the QGCs with the internet connection will have the latest board information Json file as well.

### Existing Use-Case
Ardupilot already does that (and QGC supports this Server-fetch based board information update) by this manifest link: https://firmware.ardupilot.org/manifest.json

### Information inside board_information.json
This file should include generic information about the board targets, and not be limited to bootloader ID itself (which is crucial for QGC), so that we can utilize the file in many different applications.

```
{
    "board_name"        = "HUMAN_READABLE_BOARD_NAME",      (summary)
    "target_name"       = "TARGET_NAME_FOR_BUILD",          (manufacturer_board_label)
    "description"       = "DESCRIPTION",                    (description)
    "board_id"          = "BOOTLOADER_ID",
    "image_maxsize"     = "FLASH_SIZE",
    "build_variants"    = {"VARIANT_1", "VARIANT_2", ... }, (labels)
}
```
:exclamation:  If any critical information is missing, please let me know!

## Describe possible alternatives
The Board information JSON file can be loaded into board's Flash as a [component information](https://github.com/PX4/PX4-Autopilot/pull/16039) as well, but that would mean that a PX4_FMU-V5 board would need to hold all the information about ATL_MANTIS-EDU, PX4_FMU-V3, and so on, which doesn't make sense especially since we are constrained in Flash memory.

## QGC PR
https://github.com/mavlink/qgroundcontrol/pull/10343

## Additional context
Related discussion / PRs in the past
- https://github.com/PX4/PX4-Autopilot/issues/16351
- https://github.com/mavlink/qgroundcontrol/pull/9221
- Related discussion in QGC for this feature: https://github.com/mavlink/qgroundcontrol/issues/2797

junwoo091400 committed:
Initial commit on generating board information.json

- Includes general information about Boards, like bootloader id, build
variants, etc
- will be useful for different purposes including: QGC automatically
updating the bootloader id metadata

junwoo091400 committed:
Little adjustments

- Don't print verbose debug messages if not verbose

bkueng commented on Tools/generate_board_targets_json.py:
```diff
@@ -1,31 +1,45 @@
-""" Script to generate a JSON config with all build targets (for CI) """
+"""
+Script to generate a JSON config with all build targets (for CI), in 2 different versions:
+
+1. For enumerating targets: {"include":[{"target": "px4_fmu-v5x_default", "container": "nuttx-focal"}, ...]}
+
+Called by : Tools/generate_board_targets_json.py -e
+
+---
+
+2. For board information: {"include":[{"target": "px4_fmu-v5x_default", "board_id": "51"}, ...]}
+
+Called by : Tools/generate_board_targets_json.py -b
```
There's pretty much no overlap in functionality between the two, so why not add a separate script?

junwoo091400 replied:
I wanted to keep the 'excluded manufacturers' and 'excluded board types (`test`, etc)' consistent between the two cases, but yes I do feel like there's little overlap :0

Will separate it out!

bkueng reviewed (commented):
The json structure makes sense, except can you turn the `build_variants` into a list of objects, so we can add more attributes later on?
```
      "build_variants": [
        { "name": "default" },
        { "name": "rtps" },
      ]
```

Potentially we can also keep the 'manufacturer' -> 'board' nesting instead of the now flat board list.

junwoo091400 committed:
Separate board target & info generation script

junwoo091400 commented:
Applied the comments in the latest change :wink:

junwoo091400 committed:
Additional improvements in the script

- Remove unused variable
- Use excluded_labels list

bkueng commented on Tools/generate_board_info_manifest.py:
```diff
@@ -0,0 +1,142 @@
+Script to print out the board information manifest in JSON format like this:
+
+{
+    "include":
```
This will need to be updated

bkueng commented on Tools/generate_board_info_manifest.py:
```diff
@@ -0,0 +1,142 @@
+import os
+import json
+import re
+from kconfiglib import Kconfig
```
This isn't used
```suggestion
```

bkueng commented on Tools/generate_board_info_manifest.py:
```diff
@@ -0,0 +1,142 @@
+excluded_manufacturers = ['atlflight']
+
+# Labels (board variants) we will exclude
+excluded_labels = [
+    'stackcheck', 'nolockstep', 'replay', 'test',
+    'uavcanv1' # TODO: fix and enable
+    ]
```
Do you actually need to exclude any?

junwoo091400 replied:
I didn't want to include the targets that actually doesn't exist in the server as a binary (which is dependent on the script `generate_board_targets_json.py` file. Shall I rather exclude the 'excluded-label' filter on both scripts then?

bkueng reviewed (commented):
Looking good so far.

junwoo091400 committed:
[WIP] Include defconfig file (USB Autoconnect related) info

- Still not working, possibly since I am not interpreting how to use the
Kconfiglib properly.. The loaded syms doesn't seem ok (it just includes
the base Kconfig setting I believe)

junwoo091400 committed:
WIP: Made defconfig parsing work

- Had to include the defconfig configuration file when creating the
Kconfig() object

junwoo091400 commented:
Current problems to unblock:
1. **How do we deal with targets that doesn't have `defconfig` or `firmware.prototype`, or doesn't have the specific config key we are interested in (e.g. Vendor ID) in the file?**
   ![image](https://user-images.githubusercontent.com/23277211/179483313-be472d53-4ec4-4eb7-ab95-5014829a9abb.png)
2. In QGC side, to support the auto-connect feature in USB, the `boardDescriptionFallback` has been used (for example in this [PR](https://github.com/mavlink/qgroundcontrol/pull/8929/files#diff-2314d86332283f167f8f0b777ccd989f9f7ce2b6c6e9eefdb9cbd0279557072aR71)) to add support for new boards.
   1. Is this still necessary?
   2. How can the "**boardClass**" be defined in this manifest file created in PX4's side, since there are [5 different types classified in QGC Side](https://github.com/mavlink/qgroundcontrol/blob/master/src/comm/QGCSerialPortInfo.h#L28-L35)? : Pixhawk, SiK Radio, RTK GPS, Openpilot, PX4 Flow
3.

junwoo091400 committed:
Add binary_urls attribute to board information manifest

junwoo091400 commented:
## Status of the QGC PR:

You can now selecting any build variants that are supported in PX4's build system.
![image](https://user-images.githubusercontent.com/23277211/180244800-24db4420-9b22-4f56-8952-ffb936dfe282.png)

Here when I flash after selecting 'rtps' build variant, it downloads the binary and flashes it:
![image](https://user-images.githubusercontent.com/23277211/180244924-68bdb50b-5b0f-4d37-bc43-3cdf74933e8c.png)

## Limitations
1. Still need to make the auto-commit system work, so that PX4 will automatically update the Board Info Manifest JSON in QGC side when a new build configuration is added to the system
2. Still need to make auto-caching system work, like how the ParameterMetadata is stored in QGC and gets updated / cached.
3. Make sure that all the targets that are supported are actually uploaded in the server as a binary (currently some targets aren't built and uploaded, so it isn't possible to flash)

junwoo091400 committed:
Convert product/vendor ID to integer (from hex) and other

- Include all the builds for the manifest data

junwoo091400 commented:
Note, another example showing the need for this: https://github.com/PX4/PX4-Autopilot/issues/19443#issuecomment-1093387866
